### DISCUSSION TITLE
How to be more explicit when referencing Python objects in SQL

### DISCUSSION THREAD

**Discussion by davetapley** [duckdb/duckdb#9114] [Ideas]
Building on [this example](https://github.com/duckdb/duckdb-web/blob/8a62cfc752bbb018380c9cece0408b19fecc875e/docs/api/python/overview.md?plain=1#L53-L56) I'm often moving data from Python to DuckDB like this:

```py
import pandas as pd
pandas_df = pd.DataFrame({'a': [42]})
duckdb.execute('CREATE TABLE IF NOT EXISTS tbl ...')
duckdb.execute('INSERT INTO tbl BY NAME (SELECT * FROM pandas_df)')
```

But I've been caught out a couple of times where my table and my DataFrame have the same name, e.g.:

```py
people = pd.DataFrame({'a': [42]})
duckdb.execute('CREATE TABLE IF NOT EXISTS people ...')
duckdb.execute('INSERT INTO people BY NAME (SELECT * FROM people)')
```

I've only caught these through unit testing because it doesn't fail, it just inserts zero rows (presumably the query planner is picking `people` the table before `people` the Python variable).

Is there some existing way to be more explicit that I'm referencing the Python variable name and not the table name?

**Comment by l1t1** [comment]
I don't know how to ref the python variable, but you can use `main.` to qualify duckdb table
```
>>> print(duckdb.execute('SELECT * FROM main.tbl').df())
    a
0  42
```

**Comment by davetapley** [comment]
See also: 
- https://github.com/duckdb/duckdb/issues/11687

I've also found another awkward situation with a pattern I use sometimes:

```py
@dataclass
class StoreData:
    data: DataFrame

    def store(self, con: DuckDBPyConnection):
        data = self.data
        con.execute('query using data')
```

Problem is tools like pylint and pyright complain that `data` is unused :\

**Reply by Mause** [reply]
There's a register method that could be helpful?

**Reply by davetapley** [reply]
There is? That sounds like just what I want! Where is it? 😁

**Reply by Tishj** [reply]
con.register(name, object)

**Reply by Tishj** [reply]
What it does under the hood is create a temporary view by that name so it's considered before other catalog entries

We are looking to change that to push CTEs instead, but that shouldnt affect anything :)
